packages/dds/merge-tree/src/mergeTree.ts

```diff
@@ -1414,6 +1425,116 @@ export class MergeTree {
+     */
+    public getSlideOnRemoveReferenceSegmentAndOffset(pos: number, refSeq: number, clientId: number) {
+        let segoff = this.getContainingSegment(pos, refSeq, clientId);
+        if (segoff.segment && isRemovedAndAcked(segoff.segment)) {
+            // Only slide is the segment is removed and acked
+            segoff = this.getSlideToSegment(segoff.segment);
+        }
+        if (segoff.segment && isRemoved(segoff.segment)) {
+            // All positions on removed segments must have offset 0
+            segoff.offset = 0;
+        }
+        return segoff;
+    }
+
+    /**
+     * @internal - this method should only be called by client
+     */
+    public slideReference(ref: LocalReference) {
+        const segment = ref.getSegment();
+        assert(!!segment, "slideReference requires a segment");
+        if (!isRemovedAndAcked(segment)) {
+            // We only slide the reference if the segment remove has been sequenced by the server
+            return;
+        }
+        assert(!!segment.localRefs, "Ref not in the segment localRefs");
+        const removedRef = segment.localRefs.removeLocalRef(ref);
+        assert(ref === removedRef, "Ref not in the segment localRefs");
+        const newSegoff = this.getSlideToSegment(segment);
+        const newSegment = newSegoff.segment;
+        if (!newSegment) {
+            // No valid segments (all nodes removed or not yet created)
+            ref.segment = undefined;
+            ref.offset = 0;
+            return;
+        }
+        if (!newSegment.localRefs) {
+            newSegment.localRefs = new LocalReferenceCollection(newSegment);
+        }
+        ref.segment = newSegment;
+        ref.offset = newSegoff.offset;
+        newSegment.localRefs.addLocalRef(ref);
+        // TODO is it required to update the path lengths?
+        this.blockUpdatePathLengths(newSegment.parent, TreeMaintenanceSequenceNumber,
+            LocalClientId);
+    }
+
+    /**
+     * @internal - this method should only be called by BaseSegment
+     */
+    public updateSegmentRefsAfterMarkRemoved(segment: ISegment, pending: boolean) {
+        if (!segment.localRefs || segment.localRefs.empty) {
+            return;
+        }
+        const refsToSlide: LocalReference[] = [];
+        const refsToStay: LocalReference[] = [];
+        for (const lref of segment.localRefs) {
+            if (refTypeIncludesFlag(lref, ReferenceType.StayOnRemove)) {
+                refsToStay.push(lref);
+            } else if (refTypeIncludesFlag(lref, ReferenceType.SlideOnRemove)) {
+                if (pending) {
+                    refsToStay.push(lref);
+                } else {
+                    refsToSlide.push(lref);
+                }
+            }
+        }
+        // TODO:ransomr rethink implementation of keeping and sliding refs
+        // This works but is fragile and possibly slow
+        for (const ref of refsToSlide) {
```

Comment:
In what case do refs here slide to different segments? Seems like they'd all go to the same segment.

Reply:
Thanks for catching - fixed.